packages/teleterm/src/ui/DocumentCluster/ClusterResources/useServerSideResources.ts

```diff
@@ -0,0 +1,153 @@
+import { useState, useEffect } from 'react';
+import { SortType } from 'design/DataTable/types';
+import { useAsync } from 'shared/hooks/useAsync';
+import { ServerSideParams } from 'teleterm/services/tshd/types';
+import { useAppContext } from 'teleterm/ui/appContextProvider';
+import { retryWithRelogin } from 'teleterm/ui/utils';
+import { useClusterContext } from '../clusterContext';
+import { AgentFilter, AgentLabel } from 'teleport/services/agents';
+import { addAgentLabelToQuery } from 'shared/utils/addAgentLabelToQuery';
+import { ResourceKind } from 'teleport/Discover/Shared';
+
+export function useServerSideResources(resource: ResourceKind) {
+  const ctx = useAppContext();
+  const { clusterUri, documentUri } = useClusterContext();
+  const [pageIndex, setPageIndex] = useState(0);
+  const [keys, setKeys] = useState<string[]>([]);
```

Comment:
Previously, these two values were held in a `Page` object, but these two values serve different purposes now and are accessed/used at different times. the `pageIndex` is now used to (obviously) update our current page, which will trigger a fetch effect. and `keys` is used to store data returned from that effect. If they were to be kept separate, the state would get all wonky trying to update only half of itself at "opposite" times. This smooths out the experience.